cmd/podman/common/completion.go

```diff
@@ -763,3 +746,77 @@ func AutocompleteSDNotify(cmd *cobra.Command, args []string, toComplete string)
 	types := []string{"container", "conmon", "ignore"}
 	return types, cobra.ShellCompDirectiveNoFileComp
 }
+
+// AutocompletePsFilters - Autocomplete ps filter options.
+func AutocompletePsFilters(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	id, name, status, ancestor := "id=", "name=", "status=", "ancestor="
+	before, since, volume, health := "before=", "since=", "volume=", "health="
+	filters := []string{id, name, "label=", "exited=", status, ancestor, before, since, volume, health, "until="}
+	switch {
```

Comment:
I'm finding it really hard to deal with the duplication below, and the hardcoded `[stringlength:]`. I know Go has maps, does it have closures? Could you do something like:
```
    map...
       "id=": func(s string) { getContainers(s, completeIDs) },
       "name=": func(s string) { getContainers(s, completeNames) },
       "ancestor=": func(s string) { getImages(s) },
       ...
)
...
    found := false
    for each item in map:
       if strings.HasPrefix(toComplete, item.lhs) {
            strip_arg = toComplete[len(item.lhs):]
            item.func(strip_arg)
            found = true
        }
    if ! found {
         handle status, health, etc
```

Reply:
Something like that should be possible. Great idea.

Reply:
@edsantiago PTAL I think it should be way better now. Although I wish go had macros so I wouldn't have to type `func(s string) ([]string, cobra.ShellCompDirective) {}` all time.

Reply:
That is much, much nicer, thank you! (And I agree that macros would make it even better). Looking into it some more. Thank you for your patience.